src/App.php

```diff
@@ -158,36 +147,58 @@ public function dispatch(): ResponseInterface
                 $controllerName = $routeInfo[1];
                 if (isset($routeInfo[2])) {
                     foreach ($routeInfo[2] as $name => $value) {
-                        $this->request = $this->request->withAttribute($name, $value);
+                        $request = $request->withAttribute($name, $value);
                     }
                 }
                 break;
         }
         // default HTTP response
         $response = new Response(200);
 
-        $controllerName = is_array($controllerName) ?: [$controllerName];
-        foreach ($controllerName as $controller) {
-            $this->logger->debug(sprintf("Executing %s", $controller));
+        if (!is_array($controllerName)) {
+            $controllerName = [$controllerName];
+        }
+        foreach ($controllerName as $name) {
+            $this->logger->debug(sprintf("Executing %s", $name));
             try {
-                $response = $this->container
-                    ->get($controller)
-                    ->execute($this->request, $response);
+                $controller = $this->container->get($name);
+                $response = $controller->execute($request, $response);
                 if ($response instanceof HaltResponse) {
-                    $this->logger->debug(sprintf("Found HaltResponse in %s", $controller));
+                    $this->logger->debug(sprintf("Found HaltResponse in %s", $name));
                     break;
                 }
+                // Add the PSR-7 attributes to the next request
+                if ($controller instanceof AttributeInterface) {
+                    foreach ($controller->getRequestAttribute() as $key => $value) {
+                        $request = $request->withAttribute($key, $value);
+                    }
+                }
             } catch (NotFoundExceptionInterface $e) {
                 throw new ControllerException(sprintf(
                     'The controller name %s cannot be retrieved from the container',
-                    $controller
+                    $name
                 ));
             }    
         }
-        
+
+        $this->logger->info(sprintf(
+            "Response: %d", 
+            $response->getStatusCode()
+        ));
+
         $this->logger->info(sprintf("Execution time: %.3f sec", microtime(true) - $this->startTime));
         $this->logger->info(sprintf("Memory usage: %d bytes", memory_get_usage(true)));
 
         return $response;
     }
+
+    /**
+     * Returns a PSR-7 request from globals ($_GET, $_POST, $_SERVER, etc)
+     */
+    public static function buildRequestFromGlobals(): ServerRequestInterface
```

Comment:
Am I wrong or are we not using this?

Reply:
You right, this is a static function to be used in the [skeleton](https://github.com/simplemvc/skeleton). Maybe I need to write a test for it :-)=

Reply:
Yes Please!